### PR TITLE
Merges changes to the Search module from WordPress.com

### PR DESCRIPTION
**Pull request by zinigor**
#### Changes proposed in this Pull Request:
* Merges the following changes:
* r163216-wpcom
* r161259-wpcom
* r161153-wpcom
* r156457-wpcom

### CODE REVIEW

**Comment by nickdaugherty** [comment]
Some of these changes conflict with changes that already exist in this plugin, for example the error handling added on line 374, which exists already here:

https://github.com/Automattic/jetpack/blob/f0678c20f4e2c9b6cb5baacda15d0014921fab70/modules/search/class.jetpack-search.php#L414-L430

File: modules/search/class.jetpack-search.php [hunk 1]
```diff
@@ -370,7 +370,12 @@ public function search( array $es_args ) {
 		}
 
 		$response_code = wp_remote_retrieve_response_code( $request );
-		$response      = json_decode( wp_remote_retrieve_body( $request ), true );
+
+		if ( ! $response_code || $response_code < 200 || $response_code >= 300 ) {
```
Comment (nickdaugherty) [inline modules/search/class.jetpack-search.php]: This check has been moved down in JP: https://github.com/Automattic/jetpack/blob/f0678c20f4e2c9b6cb5baacda15d0014921fab70/modules/search/class.jetpack-search.php#L414-L430

File: modules/search/class.jetpack-search.php [hunk 2]
```diff
@@ -722,6 +731,16 @@ public function get_es_wp_query_post_type_for_query( WP_Query $query ) {
 		return $sanitized_post_types;
 	}
 
+	/**
+	 * Initialze widgets for the Search module
+	 *
+	 * @module search
+	 */
+	public function action__widgets_init() {
```
Comment (nickdaugherty) [inline modules/search/class.jetpack-search.php]: This block should be removed - the widget is loaded elsewhere in JP.

File: modules/search/class.jetpack-search.php [hunk 3]
```diff
@@ -1119,6 +1138,15 @@ public function convert_wp_es_to_es_args( array $args ) {
 			unset( $es_query_args['sort'] );
 		}
 
+		if ( ! empty( $filters ) && is_array( $filters ) ) {
```
Comment (nickdaugherty) [inline modules/search/class.jetpack-search.php]: Hmm, not sure about this block - needs investigation.

**Comment by zinigor** [comment]
Thanks for the review, Nick! I have removed the duplicate check and unneeded block. As for the condition, it was added with r156256-wpcom. Looks kinda redundant because it should at least be an empty array by this point, but for code sync purposes I'd say it's totally fine.

File: modules/search/class.jetpack-search.php (hunk 3 above)
Comment (nickdaugherty) [reply modules/search/class.jetpack-search.php]: Yeah this block is wrong - it was removed in https://github.com/Automattic/jetpack/commit/84c39273984a02f66a84127f9d72fad222ade929#diff-47deebe9b1fd7090caed8291af235f54 after refactoring

**Comment by zinigor** [comment]
@nickdaugherty I have removed the block that popped back in after the merge, thanks for catching that! Now this PR looks rather harmless.

**Pull request merged**